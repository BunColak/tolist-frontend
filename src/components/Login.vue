<template>
  <div>
    <div class="fixed w-full h-full bg-gray-800 top-0 left-0 opacity-70"></div>
    <div
      class="fixed lg:w-1/2 lg:h-2/3 p-8 bg-white text-black opacity-100 top-1/2 left-1/2 transform -translate-x-1/2 -translate-y-1/2 rounded"
    >
      <form class="lg:w-2/5 mx-auto" @submit.prevent="onLoginSubmit">
        <h3 class="text-4xl text-center">Log In</h3>
        <ul class="mt-8 ml-4 text-red-700 list-disc" v-if="errorsMessages">
          <li v-for="error in errorsMessages" :key="error">{{ error }}</li>
        </ul>
        <div class="mt-4 flex flex-col">
          <div class="input-group">
            <label for="username"> Username </label>
            <input
              v-model="username"
              name="username"
              class="login-input"
              type="text"
            />
          </div>
          <div class="input-group">
            <label for="username"> Password </label>
            <input
              v-model="password"
              name="password"
              class="login-input"
              type="password"
            />
          </div>
          <button type="submit" class="button" @click="onLoginSubmit">
            Continue
          </button>
        </div>
        <div class="flex mt-8 items-center justify-center">
          <div class="h-px w-1/3 bg-gray-400" />
          <span class="mx-8">OR</span>
          <div class="h-px w-1/3 bg-gray-400" />
        </div>
      </form>
    </div>
  </div>
</template>

<script>
import LoginMutation from '../graphql/LoginMutation.gql'
import { onLogin } from '../vue-apollo'

export default {
  name: 'Login',
  data () {
    return {
      username: '',
      password: '',
      submitting: false,
      errorsMessages: []
    }
  },
  methods: {
    async onLoginSubmit () {
      this.submitting = true
      try {
        const results = await this.$apollo.mutate({
          mutation: LoginMutation,
          variables: {
            username: this.username,
            password: this.password
          }
        })
        const token = results.data.login.token
        onLogin(token)
        this.$root.token = token
      } catch (e) {
        if (e.graphQLErrors) {
          this.errorsMessages = e.graphQLErrors.map((er) => er.message)
        } else {
          console.log(e)
        }
      }
      return null
    }
  }
}
</script>

<style scoped>
.input-group {
  @apply flex flex-col mt-4;
}

.login-input {
  @apply border-2 rounded py-2 px-4 mt-2 focus:ring-indigo-300 focus:ring-2;
}
</style>
