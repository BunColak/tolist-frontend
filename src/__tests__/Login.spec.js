import { shallowMount, createLocalVue } from '@vue/test-utils'
import Login from '@/components/Login'

const localVue = createLocalVue()

describe('Login', () => {
  it('should render without problems', () => {
    const wrapper = shallowMount(Login)
    expect(wrapper).toMatchSnapshot()
  })

  it('should request login on form submit', () => {
    const mutate = jest.fn()
    mutate.mockResolvedValue({
      data: {
        login: {
          token: 'test'
        }
      }
    })
    const wrapper = shallowMount(Login, {
      localVue,
      mocks: {
        $apollo: {
          mutate
        }
      }
    })

    const variables = {
      username: 'test',
      password: 'test'
    }

    wrapper.setData(variables)

    wrapper.find('form').trigger('submit')

    expect(mutate).toBeCalledWith(
      expect.objectContaining({
        variables
      })
    )
  })

  it('should display errors in a list if there is an error', async () => {
    const mutate = jest.fn()
    const errors = ['Error 1', 'Error 2']
    mutate.mockRejectedValueOnce({
      graphQLErrors: errors.map(e => ({ message: e }))
    })

    const wrapper = shallowMount(Login, {
      localVue,
      mocks: {
        $apollo: {
          mutate
        }
      }
    })

    wrapper.find('form').trigger('submit')
    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()
    expect(wrapper.find('ul').exists()).toBeTruthy()
    expect(wrapper.findAll('li').length).toEqual(2)
  })

  it('should set token in root on login', async () => {
    const mutate = jest.fn()
    const mockToken = 'testToken'
    const Parent = {
      data () {
        return {
          token: ''
        }
      }
    }

    mutate.mockResolvedValue({
      data: {
        login: {
          token: mockToken
        }
      }
    })
    const wrapper = shallowMount(Login, {
      localVue,
      mocks: {
        $apollo: {
          mutate
        }
      },
      parentComponent: Parent
    })

    const variables = {
      username: 'test',
      password: 'test'
    }

    wrapper.setData(variables)
    wrapper.find('form').trigger('submit')

    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()

    expect(wrapper.vm.$root.token).toEqual(mockToken)
  })
})
