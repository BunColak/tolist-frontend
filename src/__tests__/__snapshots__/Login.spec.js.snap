// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render without problems 1`] = `
<div>
  <div class="fixed w-full h-full bg-gray-800 top-0 left-0 opacity-70"></div>
  <div class="fixed w-1/2 h-2/3 p-8 bg-white text-black opacity-100 top-1/2 left-1/2 transform -translate-x-1/2 -translate-y-1/2 rounded">
    <form class="w-2/5 mx-auto">
      <h3 class="text-4xl text-center">Log In</h3>
      <ul class="mt-8 ml-4 text-red-700 list-disc"></ul>
      <div class="mt-4 flex flex-col">
        <div class="input-group"><label for="username"> Username </label> <input name="username" type="text" class="login-input"></div>
        <div class="input-group"><label for="username"> Password </label> <input name="password" type="password" class="login-input"></div> <button type="submit" class="button">
          Continue
        </button>
      </div>
      <div class="flex mt-8 items-center justify-center">
        <div class="h-px w-1/3 bg-gray-400"></div> <span class="mx-8">OR</span>
        <div class="h-px w-1/3 bg-gray-400"></div>
      </div>
    </form>
  </div>
</div>
`;
